### DISCUSSION TITLE
Fedimint atomic broadcast abstraction based on Aleph BFT

### DISCUSSION THREAD

**Discussion by joschisan** [fedimint/fedimint#2355] [General]
In an effort to switch to Aleph BFT kitman and I have designed an atomic broadcast API to encapsulate aleph_bft. Furthermore, the broadcast abstraction provides additional functionality on top of aleph bft and would be implemented in a new crate fedimint-atomic-broadcast that would depend on fedimint-core only while we would have fedimint-server depend on fedimint-atomic-broadcast. The atomic broadcast would handle session switching while hiding the concept of sessions from the user; furthermore, the recovery is handled internally by the atomic broadcast and is also hidden from the user. All nodes can crash anytime without creating inconsistent state or requiring guardian intervention to restart. All communication for consensus, session switching and recovery is handled through an unreliable network layer as the broadcast implements all necessary retry logic; this should simplify the network layer significantly.

We would now invite everybody to take a look at this API to raise any open questions and possibly request further functionality needed for fedimint. As the design is very different from hbbft we believe there will be a lot to discuss; therefore, kitman and I propose this API as a topic for an upcoming deep dive so we can update this discussion accordingly.

```rust
// the message type the atomic broadcast uses to establish consensus. 
pub struct Message{};

// indicates to which peer a given message needs to be send
pub enum Recipient {
    Everyone,
    Peer(PeerId)
}

pub fn start(
    peer_count: usize,
    peer_id: PeerId,
    public_key_set: threshold_crypto::PublicKeySet,
    secret_key_share: threshold_crypto::SecretKeyShare,
    // all data necessary to rejoin is backed up here as files; no database required.
    backup_dir: std::path::Path,
    // this is the number of items that have already been processed by the system. We need to maintain
    // a counter for this via database transactions. Hence we need a database transaction for every item
    skip_items: u64,
) -> (
    // the channel is bounded. items may be dropped as the clients implement retry logic
    tokio::sync::mpsc::Sender<ConsensusItem>,
    // this is the ordered stream of finalized items. It is guaranteed to be identical for all peers.
    // the peer id corresponds to the peer which submitted the item. ConsensusItem::Transactions 
    // is deduplicated within a block (not within the entire consensus history) 
    tokio::sync::mpsc::Receiver<(ConsensusItem, PeerId)>,
    // the channel is bounded; messages may be dropped as aleph bft assumes an unreliable network layer
    tokio::sync::mpsc::Sender<Message>,
    // only try to send them once; all necessary retry logic is handled by aleph bft
    tokio::sync::mpsc::Receiver<(Message, Recipient)>,
    // this struct allows access to the broadcasts history of ordered items via HistoryStreams
    History
) {}

struct Block {
    // index of the block
    block_index: u64,
    // number of items contained in blocks of smaller index
    item_index: u64,
    // blocks may contain not items just as bitcoin blocks;
    // the peer id indicates which peer submitted the item to consensus
    items: Vec<(ConsensusItem, PeerId)>,
    // a threshold signature for the message block_index || item_index || root where root is 
    // the root hash of the merkle tree with items and corresponding peer ids as leafs.
    signature: Signature
}

struct History {}

impl History {
    // returns a stream of all completed and threshold signed blocks starting 
    // with the block that contains the item with given index. As the block 
    // corresponding to the current session is not yet signed it is not returned 
    // by the stream
    fn get_stream(index: u64) -> HistoryStream {}
}

struct HistoryStream{}

impl HistoryStream{
    // on every call the next block is read from file
    async fn next() -> Option<Block> {}
}
```

**Comment by elsirion** [comment]
Looks really good overall, I'd kinda like to explore how this would translate into traits a little more to see how we could make the consensus plug'n'play. I.e. which types differ for different consensus algos (e.g. config, messages sent, …), these would become associated types of a `AtomicBroadcast` trait. Then there would be a `IAtomicBroadcast` trait that uses the `Dyn*` equivalents for these associated types (e.g. look at [`ServerModule`](https://github.com/fedimint/fedimint/blob/3c271c6e5ec6df2d1092499526e6926ef7765fc5/fedimint-core/src/core/server.rs#L52) and [`IServerModule`](https://github.com/fedimint/fedimint/blob/3c271c6e5ec6df2d1092499526e6926ef7765fc5/fedimint-core/src/core/server.rs#L52) and the [translation between the two](https://github.com/fedimint/fedimint/blob/3c271c6e5ec6df2d1092499526e6926ef7765fc5/fedimint-core/src/core/server.rs#L193)).

Some other thoughts:
* We probably want to get away from `threshold_crypto` when switching away from `hbbft` (we only use it because it's already a dependency, but just as `hbbft` it isn't maintained)
* The `Message` type should probably be an associated type of a `AtomicBroadcast` trait
* I'm not sure if returning certain receivers/senders is the right move, could we instead give the constructor a network object?
* In any way, I think it would be nice to group the returned senders/receivers into structs (e.g. consensus, network, …)
* Why do sent messages have a recipient but received messages no sender? Is the sender encapsulated in the message but the recipient isn't?
* Might it be prudent to have each block commit to its predecessor?

**Reply by joschisan** [reply]
* we will probably do a trait for testing at some point after we have found a stable interface
* yes we could pass in a network object similar to the aleph bft API. why are you wary of returning senders/receivers?
* do you mean you want to group them into structs just to return them or actually have a consensus/network struct which hides the underlaying sender and receiver - in the latter case this seems like a lot of boilerplate as we would expose most of the functionality of the channels; furthermore; we may want to submit items from several different tasks but can only have one task processing the ordered stream.
* yes, the sender is encapsulated in the message but the recipient is not
* we see no advantage to have each block commit to it's predecessor; though this might save space on signatures this would be negligible as blocks are quite big and only add complexity

Generally the most important thing to discuss right now seems to be the Blocks and HistoryStreams as especially the block struct will be saved to disk by the broadcast and is difficult to change later on:
* what data needs to be signed in order to be verified by clients
* do we use a header struct that contains just the signed data; so only a root hash and no items
*  do we need inclusion proofs
* is it okay that we do not return the unsigned partial block of the ongoing session from the HistoryStream
